src/Npgsql/TypeHandlers/CompositeHandlers/MappedCompositeHandler`1.cs

```diff
@@ -0,0 +1,152 @@
+using System;
+using System.Linq;
+using System.Reflection;
+using System.Threading.Tasks;
+using Npgsql.BackendMessages;
+using Npgsql.PostgresTypes;
+using Npgsql.TypeHandling;
+using Npgsql.TypeMapping;
+using NpgsqlTypes;
+
+namespace Npgsql.TypeHandlers.CompositeHandlers
+{
+    class MappedCompositeHandler<T> : NpgsqlTypeHandler<T>, IMappedCompositeHandler
+        where T : new()
+    {
+        readonly CompositeMemberInfo<T>[] _members;
+
+        public Type CompositeType => typeof(T);
+
+        MappedCompositeHandler(PostgresCompositeType postgresType, CompositeMemberInfo<T>[] members)
+        {
+            PostgresType = postgresType;
+            _members = members;
+        }
+
+        public override async ValueTask<T> Read(NpgsqlReadBuffer buffer, int length, bool async, FieldDescription fieldDescription = null)
+        {
+            await buffer.Ensure(sizeof(int), async);
+
+            var fieldCount = buffer.ReadInt32();
+            if (fieldCount != _members.Length)
+                throw new InvalidOperationException($"pg_attributes contains {_members.Length} fields for type {PgDisplayName}, but {fieldCount} fields were received.");
+
+            var composite = new T();
```

Comment:
Yep, will do, but in a separate PR.

Reply:
Re constructor mapping, let's not do this right away - nobody has ever asked for it and there's lots of other work :) We can open an issue so that people can vote though.

IMHO, Npgsql's composite doesn't pretend to be a full-fledged mapping technology like EF Core, and composite types are unlikely to be used by users in the way EF Core entities are used. So I'd rather hear some user requests for this (and understand the usecase) before jumping too quickly with it...

Reply:
> Re constructor mapping, let's not do this right away - nobody has ever asked for it and there's lots of other work :) We can open an issue so that people can vote though.

I remember I saw an issue about it from F# community. Anyway don't forget about immutability, for structures it's only one good way as you know.

Reply:
Yeah. I think it could make sense and I'm not against it, just saying we should wait until this feature is requested, that's all (there's lots of stuff to do).